services/azure-devops/azure-devops-coverage.service.js

```diff
@@ -0,0 +1,168 @@
+<p>
+  Your badge will then have the form:
+  <code>https://img.shields.io/azure-devops/coverage/ORGANIZATION/PROJECT/DEFINITION_ID.svg</code>.
+</p>
+<p>
+  Optionally, you can specify a named branch:
+  <code>https://img.shields.io/azure-devops/coverage/ORGANIZATION/PROJECT/DEFINITION_ID/NAMED_BRANCH.svg</code>.
+</p>
+`
+const {
+  coveragePercentage: coveragePercentageColor,
+} = require('../../lib/color-formatters')
+
+const latestBuildSchema = Joi.object({
+  count: Joi.number().required(),
+  value: Joi.array()
+    .items(
+      Joi.object({
+        id: Joi.number().required(),
+      })
+    )
+    .required(),
+}).required()
+
+const buildCodeCoverageSchema = Joi.object({
+  coverageData: Joi.array()
+    .items(
+      Joi.object({
+        coverageStats: Joi.array()
+          .items(
+            Joi.object({
+              label: Joi.string().required(),
+              total: Joi.number().required(),
+              covered: Joi.number().required(),
+            })
+          )
+          .min(1)
+          .required(),
+      })
+    )
+    .required(),
+}).required()
+
+module.exports = class AzureDevOpsCoverage extends BaseJsonService {
+  static render({ coverage }) {
+    return {
+      message: `${coverage.toFixed(0)}%`,
+      color: coveragePercentageColor(coverage),
+    }
+  }
+
+  static get defaultBadgeData() {
+    return { label: 'coverage' }
+  }
+
+  static get category() {
+    return 'build'
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Azure DevOps coverage',
+        urlPattern:
+          'azure-devops/coverage/:organization/:project/:definitionId',
+        staticExample: this.render({ coverage: 100 }),
+        exampleUrl: 'azure-devops/coverage/swellaby/opensource/25',
+        keywords: ['vso', 'vsts', 'azure-devops'],
+        documentation,
+      },
+      {
+        title: 'Azure DevOps coverage (branch)',
+        urlPattern:
+          'azure-devops/coverage/:organization/:project/:definitionId/:branch',
+        staticExample: this.render({ coverage: 100 }),
+        exampleUrl: 'azure-devops/coverage/swellaby/opensource/25/master',
+        keywords: ['vso', 'vsts', 'azure-devops'],
+        documentation,
+      },
+    ]
+  }
+
+  static get route() {
+    return {
+      base: '',
+      format: '(?:azure-devops|vso)/coverage/([^/]+)/([^/]+)/([^/]+)(?:/(.+))?',
+      capture: ['organization', 'project', 'definitionId', 'branch'],
+    }
+  }
+
+  async fetch({ url, options, schema }) {
+    return this._requestJson({
+      url,
+      options,
+      schema,
+      errorMessages: {
+        404: 'build pipeline or coverage not found',
+      },
+    })
+  }
+
+  async getLatestBuildId(organization, project, definitionId, branch, options) {
+    let url = `https://dev.azure.com/${organization}/${project}/_apis/build/builds?definitions=${definitionId}&$top=1&api-version=5.0-preview.4`
```

Comment:
Could you pass this to `request` through a `options.qs` object instead? That way the parameters will be correctly url-encoded.

I realize `options` is being generated through a helper. Since the options will then depend on the request, perhaps it would be clearer for the helper to return either the `headers` object or the auth header. Then this could be e.g.:

```js
options: {
  qs: { ... }
  headers: { Authorization: createAuthHeader() }
}
```

Reply:
Good idea